src/Components/Components/src/Rendering/HtmlRenderer.cs

```diff
@@ -189,6 +193,29 @@ private int RenderChildren(List<string> result, ArrayRange<RenderTreeFrame> fram
                 return position;
             }
 
+            // We need to de-duplicate attributes when rendering to HTML. We allow *last-attribute-wins* when writing
+            // Blazor code, but HTML is the opposite.
+            //
+            // To implement this we do two passes through the attributes:
+            // - On the first pass, record the index of the *last* occurrence of the attribute.
+            // - On the second pass, write each attribute *if* it's the last occurrence.
+            //
+            // This process is done by the compiler for markup blocks (fully static HTML), so we don't need to repeat
+            // it there.
+            _currentAttributes.Clear();
```

Comment:
@SteveSandersonMS @javiercn - to implement "last attribute wins" behaviour we either need to do this:
- inside the `RenderTreeBuilder` by having it buffer all attributes
- inside the renderer implementations

I'm interested in ideas about which one we should do and why. I started with the *inside the renderer* approach because it already "just works" except for the static html case.

However, this is precedent for the `RenderTreeBuilder` to be *smart*. It already knows to omit bool false values.

Reply:
If we can do it cheaply enough inside the builder that would keep things maximally consistent, but I’d hope it doesn’t add cost in the non-splatting cases. For example we could go into buffered mode only once we see the first add-splatted-attributes call, if it’s not too complex. 

Otherwise I’m unsure how the diff algo will handle it. It doesn’t expect to see duplicate attributes, and since it uses a dictionary-based merge for attributes sometimes, that will need to be checked carefully to ensure things get overwritten in the correct order.

Reply:
Hmm. There's actually a problem with this approach in general. Currently the RTB omits frames for things like `bool false` attributes. The technique used here is going to have subtle bugs unless all calls to `AddAttribute` result in a frame. 

```
<div @attributes disable="@disableMyDiv">...</div>
```

We want the effect of `disable="@disableMyDiv"` to supersede whatever is in `@myAttribtues`.

I'll try to the opposite approach and see what happens. I have an idea of how to do without any additional data structures.

Reply:
@SteveSandersonMS - please take a look at how this is handled inside the RTB now.